src/Checkout/ExpressCheckout/SalesChannel/ExpressShippingCallbackRoute.php

```diff
@@ -0,0 +1,101 @@
+    ) {
+    }
+
+    public function getDecorated(): AbstractExpressShippingCallbackRoute
+    {
+        throw new DecorationPatternException(self::class);
+    }
+
+    #[OA\Post(
+        path: '/paypal/express/shipping-callback',
+        operationId: 'handlePayPalExpressShippingCallback',
+        description: 'Handles PayPal shipping address change callbacks and returns updated cart pricing',
+        requestBody: new OA\RequestBody(content: new OA\JsonContent(properties: [
+            new OA\Property(property: 'id', description: 'PayPal order ID', type: 'string'),
+            new OA\Property(property: 'shipping_address', type: 'object'),
+            new OA\Property(property: 'purchase_units', type: 'array', items: new OA\Items(type: 'object')),
+        ])),
+        tags: ['Store API', 'PayPal'],
+        responses: [new OA\Response(
+            response: Response::HTTP_OK,
+            description: 'Updated purchase units with recalculated prices',
+            content: new OA\JsonContent(properties: [
+                new OA\Property(property: 'purchase_units', type: 'array', items: new OA\Items(type: 'object')),
+            ])
+        )],
+    )]
+    #[Route(path: '/store-api/paypal/express/shipping-callback', name: 'store-api.paypal.express.shipping_callback', methods: ['POST'], defaults: ['csrf_protected' => false])]
+    public function handleCallback(Request $request, SalesChannelContext $salesChannelContext): JsonResponse
+    {
+        try {
+            $this->logger->info('PayPal shipping callback received', [
+                'paypalOrderId' => $request->request->get('id'),
+                'shippingAddress' => $request->request->get('shipping_address'),
+            ]);
+
+            $paypalOrderId = $request->request->get('id');
+            $shippingAddress = $request->request->get('shipping_address');
+
+            if (!\is_string($paypalOrderId) || !\is_array($shippingAddress)) {
+                $this->logger->error('Invalid callback payload', [
+                    'payload' => $request->request->all(),
+                ]);
+
+                return new JsonResponse(['error' => 'Invalid payload'], Response::HTTP_BAD_REQUEST);
+            }
+
+            $updatedPurchaseUnits = $this->shippingCallbackService->recalculateCart(
+                $paypalOrderId,
+                $shippingAddress,
+                $salesChannelContext
+            );
+
+            $this->logger->info('PayPal shipping callback processed successfully', [
+                'paypalOrderId' => $paypalOrderId,
+                'updatedAmount' => $updatedPurchaseUnits[0]['amount']['value'] ?? null,
+            ]);
+
+            return new JsonResponse(['purchase_units' => $updatedPurchaseUnits]);
+        } catch (\Throwable $e) {
+            $this->logger->error('PayPal shipping callback failed', [
+                'error' => $e->getMessage(),
+                'trace' => $e->getTraceAsString(),
+            ]);
```

Comment:
Our `IntrospectionProcessor` will care about the rest:
```suggestion
            $this->logger->error('Shipping callback failed', ['exception' => $e]);
```